pkg/kubelet/watchdog/watchdog_linux.go

```diff
@@ -0,0 +1,158 @@
+}
+
+// DefaultWatchdogClient implements the WatchdogClient interface using the actual systemd daemon functions.
+type DefaultWatchdogClient struct{}
+
+var _ WatchdogClient = &DefaultWatchdogClient{}
+
+func (d *DefaultWatchdogClient) SdWatchdogEnabled(unsetEnvironment bool) (time.Duration, error) {
+	return daemon.SdWatchdogEnabled(unsetEnvironment)
+}
+
+func (d *DefaultWatchdogClient) SdNotify(unsetEnvironment bool) (bool, error) {
+	return daemon.SdNotify(unsetEnvironment, daemon.SdNotifyWatchdog)
+}
+
+// Option defines optional parameters for initializing the healthChecker
+// structure.
+type Option func(*healthChecker)
+
+func WithWatchdogClient(watchdog WatchdogClient) Option {
+	return func(hc *healthChecker) {
+		hc.watchdog = watchdog
+	}
+}
+
+type healthChecker struct {
+	checkers     []healthz.HealthChecker
+	retryBackoff wait.Backoff
+	interval     time.Duration
+	watchdog     WatchdogClient
+}
+
+var _ HealthChecker = &healthChecker{}
+
+const notifyInterval = time.Nanosecond
+
+// NewHealthChecker creates a new HealthChecker instance.
+// This function initializes the health checker and configures its behavior based on the status of the systemd watchdog.
+// If the watchdog is not enabled, the function returns an error.
+func NewHealthChecker(syncLoop syncLoopHealthChecker, opts ...Option) (HealthChecker, error) {
+	hc := &healthChecker{
+		watchdog: &DefaultWatchdogClient{},
+	}
+	for _, o := range opts {
+		o(hc)
+	}
+
+	// get watchdog information
+	watchdogVal, err := hc.watchdog.SdWatchdogEnabled(false)
+	if err != nil {
+		// Failed to get watchdog configuration information.
+		// This occurs when we want to start the watchdog but the configuration is incorrect,
+		// for example, the time is not configured correctly.
+		return nil, fmt.Errorf("configure watchdog: %w", err)
+	}
+	if watchdogVal == 0 {
+		klog.InfoS("Systemd watchdog is not enabled")
+		return &healthChecker{}, nil
+	}
+	if watchdogVal <= notifyInterval {
+		return nil, fmt.Errorf("configure watchdog timeout too small: %v", watchdogVal)
+	}
+
+	// The health checks performed by checkers are the same as those for "/healthz".
+	checkers := []healthz.HealthChecker{
+		healthz.PingHealthz,
+		healthz.LogHealthz,
+		healthz.NamedCheck("syncloop", syncLoop.SyncLoopHealthCheck),
+	}
+	retryBackoff := wait.Backoff{
+		Duration: time.Second,
+		Factor:   2.0,
+		Jitter:   0.1,
+		Steps:    2,
+	}
+	hc.checkers = checkers
+	hc.retryBackoff = retryBackoff
+	hc.interval = watchdogVal / 2
+
+	return hc, nil
+}
+
+func (hc *healthChecker) Start() {
+	if hc.interval <= 0 {
+		klog.InfoS("Systemd watchdog is not enabled or the interval is invalid, so health checking will not be started.")
+		return
+	}
+	klog.InfoS("Starting systemd watchdog with interval", "interval", hc.interval)
+
+	go wait.Forever(func() {
+		if err := hc.doCheck(); err != nil {
+			klog.ErrorS(err, "Do not notify watchdog this iteration as the kubelet is reportedly not healthy")
```

Comment:
Optionally, notify `STATUS` to report the unhealthiness.

Reply:
We can also notify `WATCHDOG=trigger` to immediately inform systemd to consider the kubelet unit unhealthy.

Reply:
What is the watchdog behavior if we send the trigger? Will it restart immediately? I feel like waiting for the interval configured by user seems to be a safer and more predictable design

Reply:
> Optionally, notify `STATUS` to report the unhealthiness (and overwrite that once the kubelet is healthy again).

This is good, as it allows we to use `systemctl status kubelet.service`to see what happened with the kubelet.

Reply:
> What is the watchdog behavior if we send the trigger? Will it restart immediately? I feel like waiting for the interval configured by user seems to be a safer and more predictable design

+1